### PR TITLE
Feat/create the engine class gui

### PR DESCRIPTION
**Pull request by mathieurobert1**
I create the Engine class.
This class is the main class of the GUI. Its method run() is the loop that listen server, listen events and display the game.

More details : #17

### CODE REVIEW

File: gui/src/main.cpp [hunk 1]
```diff
@@ -5,10 +5,30 @@
...
+{
+    std::cout << "USAGE:\t./zappy_gui -p port -h machine" << std::endl;
+}
+
+int main(int argc, char **argv)
 {
-    std::cout << "Welcome to GUI" << std::endl;
+    if (argc != 3) {
+        displayHelp();
+        return 0;
```
Comment (Marius-P1) [inline gui/src/main.cpp:22]:
```suggestion
        return 84;
```

File: gui/src/main.cpp [hunk 2]
```diff
@@ -5,10 +5,30 @@
...
+    }
+    try {
+        Gui::Network net(atoi(argv[1]), argv[2]);
```
Comment (Marius-P1) [inline gui/src/main.cpp:25]: The argument are not handled like that, you can refer yourself to the subject

File: gui/src/main.cpp [hunk 3]
```diff
@@ -5,10 +5,30 @@
...
-int main(void)
+static void displayHelp()
+{
+    std::cout << "USAGE:\t./zappy_gui -p port -h machine" << std::endl;
+}
+
+int main(int argc, char **argv)
 {
-    std::cout << "Welcome to GUI" << std::endl;
+    if (argc != 3) {
```
Comment (Njord201) [inline gui/src/main.cpp:20]: You have at least 5 arguments in the command of gui

**Review by Njord201 — CHANGES_REQUESTED** [review]
One request

**Review by Marius-P1 — CHANGES_REQUESTED** [review]
As said, you have to update how you handle your arguments

File: gui/src/Engine/Engine.cpp [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+void Gui::Engine::listenServer()
+{
+    std::string command = _network.listenServer();
+
+    if (command == "")
+        return;
+    try {
+        std::vector<std::string> arguments = _parser.parse(command);
+        //TODO: update game with arguments...
```
Comment (Njord201) [inline gui/src/Engine/Engine.cpp:32]: Create an issue

File: gui/include/Engine/Engine.hpp [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    private:
+
+        ServerParser    _parser;    // Parser class for server's command
+        Network         _network;   // Network class to connect to the server.
+
+        /**
+         * @brief Listen the server and update Engine with its commands.
+         *
+         */
+        void listenServer();
```
Comment (Njord201) [inline gui/include/Engine/Engine.hpp:54]:
```suggestion
        void listenServer(void);
```

File: gui/src/Engine/Engine.cpp [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+{
+    //while (//TODO : renderer.getIsOpen()) {
+        listenServer();
+        // TODO :
+        // handleEvents();
+        // display();
+    // }
+}
+
+void Gui::Engine::listenServer()
```
Comment (Njord201) [inline gui/src/Engine/Engine.cpp:24]:
```suggestion
void Gui::Engine::listenServer(void)
```

File: gui/include/Engine/Engine.hpp [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+         * @brief Destroy the Engine object.
+         *
+         */
+        ~Engine() = default;
+
+        /**
+         * @brief Run the engine loop.
+         *
+         */
+        void run();
```
Comment (Njord201) [inline gui/include/Engine/Engine.hpp:43]:
```suggestion
        void run(void);
```

File: gui/src/Engine/Engine.cpp [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+** Engine
+*/
+
+#include "Engine/Engine.hpp"
+
+#include <iostream>
+
+Gui::Engine::Engine(Network network) : _network(network) {}
+
+void Gui::Engine::run()
```
Comment (Njord201) [inline gui/src/Engine/Engine.cpp:14]:
```suggestion
void Gui::Engine::run(void)
```

File: gui/include/Error/Error.hpp [hunk 9]
```diff
@@ -45,5 +45,19 @@ namespace Gui {
...
                 ServerParserException(std::string message);
         };
+
+        /**
+         * @class ParseCommandLineException
+         * @brief Error class for parseCommandLine errors.
+         */
+        class ParseCommandLineException : public Error {
+            public:
+                /**
```
Comment (Njord201) [inline gui/include/Error/Error.hpp:55]:
```suggestion
            public:

                /**
```

File: gui/include/Parsing/ParseCommandLine.hpp [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+         *
+         */
+        ~ParseCommandLine() = default;
+
+        /**
+         * @brief Get the port object.
+         *
+         * @return int - port
+         */
+        int getPort();
```
Comment (Njord201) [inline gui/include/Parsing/ParseCommandLine.hpp:44]:
```suggestion
        int getPort(void);
```

File: gui/include/Parsing/ParseCommandLine.hpp [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+        /**
+         * @brief Get the hostName object.
+         *
+         * @return std::string - hostName
+         */
+        std::string getHostName();
```
Comment (Njord201) [inline gui/include/Parsing/ParseCommandLine.hpp:51]:
```suggestion
        std::string getHostName(void);
```

File: gui/src/Parsing/ParseCommandLine.cpp [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+#include <string.h>
+#include <iostream>
+
+Gui::ParseCommandLine::ParseCommandLine(int argc, char **argv)
+{
+    bool isPort = false;
+    bool isHostname = false;
+
+    if (argc != 5)
+        throw Errors::ParseCommandLineException("USAGE:\t./zappy_gui -p port -h machine");
```
Comment (Njord201) [inline gui/src/Parsing/ParseCommandLine.cpp:20]: Create a MACRO for the -h.

File: gui/src/Parsing/ParseCommandLine.cpp [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+    for (int i = 1; i <= 3; i += 2) {
+        if (strcmp("-p", argv[i]) == 0 && i + 1 < argc) {
+            _port = atoi(argv[i + 1]);
+            isPort = true;
+        } else if (strcmp("-h", argv[i]) == 0 && i + 1 < argc) {
+            _hostName = argv[i + 1];
+            isHostname = true;
+        } else
+            throw Errors::ParseCommandLineException("USAGE:\t./zappy_gui -p port -h machine");
```
Comment (Njord201) [inline gui/src/Parsing/ParseCommandLine.cpp:30]: Create a MACRO.

File: gui/src/Parsing/ParseCommandLine.cpp [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+{
+    bool isPort = false;
+    bool isHostname = false;
+
+    if (argc != 5)
+        throw Errors::ParseCommandLineException("USAGE:\t./zappy_gui -p port -h machine");
+
+    for (int i = 1; i <= 3; i += 2) {
+        if (strcmp("-p", argv[i]) == 0 && i + 1 < argc) {
+            _port = atoi(argv[i + 1]);
```
Comment (Njord201) [inline gui/src/Parsing/ParseCommandLine.cpp:24]: Check of the string is a number before use atoi (atoi have not error handling). You can use std::stoul if you want (with a try, catch)

File: gui/src/Parsing/ParseCommandLine.cpp [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+            isPort = true;
+        } else if (strcmp("-h", argv[i]) == 0 && i + 1 < argc) {
+            _hostName = argv[i + 1];
+            isHostname = true;
+        } else
+            throw Errors::ParseCommandLineException("USAGE:\t./zappy_gui -p port -h machine");
+    }
+    if (!isPort || !isHostname) {
+        std::cout << isPort << isHostname << std::endl;
+        throw Errors::ParseCommandLineException("USAGE:\t./zappy_gui -p port -h machine");
```
Comment (Njord201) [inline gui/src/Parsing/ParseCommandLine.cpp:34]: Use a MACRO.

File: gui/tests/Parsing/TestParseCommandLine.cpp [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+    av[5] = NULL;
+
+    bool test = false;
+    try {
+        Gui::ParseCommandLine parser(ac, av);
+    } catch (const std::exception &error) {
+        test = true;
+    }
+    cr_assert_eq(test, true);
+}
```
Comment (Njord201) [inline gui/tests/Parsing/TestParseCommandLine.cpp:133]:
```suggestion
}

```

**Review by Njord201 — CHANGES_REQUESTED** [review]
Some request

File: gui/include/Parsing/ParseCommandLine.hpp [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+
+    /**
+     * @brief Main GUI class.
+     *
+     */
+    class ParseCommandLine;
+};
+
+class Gui::ParseCommandLine {
+
```
Comment (Marius-P1) [inline gui/include/Parsing/ParseCommandLine.hpp:22]:
```suggestion
```

File: gui/src/Parsing/ParseCommandLine.cpp [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+    }
```
Comment (Marius-P1) [inline gui/src/Parsing/ParseCommandLine.cpp:35]: Maybe you can do that in a separate function

**Review by Marius-P1 — COMMENTED** [review]
A little change

**Review by Njord201 — APPROVED** [review]

**Review by Marius-P1 — APPROVED** [review]
:ok_hand:

**Review by Thomaltarix — APPROVED** [review]
Great :+1:

**Pull request merged**